finality-aleph/src/sync/substrate/chain_status.rs

```diff
@@ -0,0 +1,147 @@
+use std::{
+    fmt::{Display, Error as FmtError, Formatter},
+    marker::PhantomData,
+};
+
+use aleph_primitives::{BlockNumber, ALEPH_ENGINE_ID};
+use log::warn;
+use sp_blockchain::Backend;
+use sp_runtime::{
+    generic::BlockId as SubstrateBlockId,
+    traits::{Block as BlockT, Header as SubstrateHeader},
+};
+
+use crate::{
+    justification::backwards_compatible_decode,
+    sync::{substrate::Justification, BlockStatus, ChainStatus, Header, LOG_TARGET},
+    AlephJustification,
+};
+
+/// What can go wrong when checking chain status
+#[derive(Debug)]
+pub enum Error<B: BlockT> {
+    MissingHash(B::Hash),
+    MissingJustification(B::Hash),
+}
+
+impl<B: BlockT> Display for Error<B> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        use Error::*;
+        match self {
+            MissingHash(hash) => {
+                write!(f, "no block for hash {:?}", hash)
+            }
+            MissingJustification(hash) => {
+                write!(
+                    f,
+                    "no justification for finalized block with hash {:?}",
+                    hash
+                )
+            }
+        }
+    }
+}
+
+/// Substrate implementation of ChainStatus trait
+struct SubstrateChainStatus<B, BE>
```

Comment:
why not `pub` or `pub(super)`?

Reply:
Should be pub, just forgot about it, as it is not used anywhere right now.